### ISSUE TITLE
Bookmark feature to save twts to a bookmarks.txt file

### ISSUE THREAD

**Issue by sorenpeter** [jointwt/twtxt#336]
A nice feature of micro.blog, RSS-readers and e-mail clients is the option to star or flag messages.

In the UI there would be a icon with a star or flag, that when clicked would copy the current twt to a bookmarks.txt in the following syntax, where <nick url> is of the author of the twt:

`<nick url> 2016-02-04T13:30:00+01:00 twtxt twtxt twtxt`

These should not be public as default, but they could be.

**Comment by prologic** [comment]
Nice idea! I'm glad you're asking for this. I would change the implemtnation slightly however:

> When a user clicks on the "Bookmark" icon next to a Twt a backend API will append that Twt's URI (_identified by its permalink_) and add it to a user's `bookmarks.txt` file next to their feed.

Additionally we _could_ choose to expose this and allow the user to choose whether or not they want to expose their bookmarks publically.

**Comment by prologic** [comment]
So for example a sample `bookmarks.txt` file for me _could_ be:

```
2021-01-20T09:43+10:00.   https://twtxt.net/twt/ucdqska
```

**Comment by sorenpeter** [comment]
It would be easier to implement, but it would only work within the twt.social world and not for other clients.
This way you would have to click all the links to access the information, which kinda defies my idea with a bookmark list, that would show you the original content;)

**Comment by prologic** [comment]
> It would be easier to implement, but it would only work within the twt.social world and not for other clients.
> This way you would have to click all the links to access the information, which kinda defies my idea with a bookmark list, that would show you the original content;)

You are correct. This isn't really _strictly_ a bookmark then I guess. Perhaps this needs some more thought. This is more or less you just copying a Twt someone else wrote into a file because you want to preserve it for some reason.

This _could_ work. @der-lyse @JonLundy any thoughts on this?

**Comment by prologic** [comment]
Okay, I had toms eimt to think about this on the _shitter_ 🚽 🤣 ...

Here's the thing. Twt Hashes are a way to content address a user's post, no matter how they creates it. It works across __all__ clients. We _shoudl_ take advantage of that. Having a list of hashes as bookmarks does not preclude any other client from being able to do the same thing, in fact they can. Most clients will fetch feeds and also cache them locally. Twet does this, I _assume_ twtnish does this too. Once cache, you have all the data you need to construct and/or compute hashes and lookup twts by their hash.

So yes. We will build this feature, but it will be a list of hashes. Any client can implement this and what we _should_ do is just spec this up as a client-agnostic feature like what we've done with some of the extensions documented at https://dev.twtxt.net/

**Comment by JonLundy** [comment]
An idea I have for future is to index a twt file. Something like a map[hash]int where it points to the seek position in the file to read a twt. 

With that a bookmark would be just a hash to look up.. though that could be more usefull if had metadata attached. Like say a searchable description topic for a group of hashes / subject threads / etc. 



Also, I love to bring it up whenever I discuss hashes. It would be awesome if it was defined as `[feedURL] + '\n' + [rawLine] + '\n'` that way tools that index hashes for a feed don't need to know anything about the twt or how to parse dates. It just hashes a block of bytes in a newline delimited file.

**Comment by JonLundy** [comment]
do we have a feature ticket for twt wiki? this topic could have some related analogs.

**Comment by prologic** [comment]
> Also, I love to bring it up whenever I discuss hashes. It would be awesome if it was defined as `[feedURL] + '\n' + [rawLine] + '\n'` that way tools that index hashes for a feed don't need to know anything about the twt or how to parse dates. It just hashes a block of bytes in a newline delimited file. 

So you want to drop the timestamp from the hash?

**Comment by prologic** [comment]
> do we have a feature ticket for twt wiki? this topic could have some related analogs.

No we don't but that's an interesting overlap. That's what a few users use those bang operators for like !resume for example

**Comment by JonLundy** [comment]
> > do we have a feature ticket for twt wiki? this topic could have some related analogs.
> 
> No we don't but that's an interesting overlap. That's what a few users use those bang operators for like !resume for example

Added. #338

**Comment by JonLundy** [comment]
> > do we have a feature ticket for twt wiki? this topic could have some related analogs.
> 
> No we don't but that's an interesting overlap. That's what a few users use those bang operators for like !resume for example

Added. #338

**Comment by JonLundy** [comment]
> So you want to drop the timestamp from the hash?

No. It's part of the raw line. So 
```
https://twtxt.net/user/prologic/twtxt.txt
2020-07-18T12:39:52Z	Hello World! 😊
```

**Comment by prologic** [comment]
Ahh I see ! Can we guarantee the hash is the same if we do this for existing twts?

**Comment by JonLundy** [comment]
If we make the tab between time and text a newline it *should* largely be the same. The non pod feeds might have trouble. If say they had tz +00:00 or include nano seconds that are all 000s

But I bet they aren't hash replied that much.

**Comment by sorenpeter** [comment]
wait, are you replies on the right issue? This is one about bookmarks, but it sounds like you are talking about conversations...

**Comment by JonLundy** [comment]
> wait, are you replies on the right issue? This is one about bookmarks, but it sounds like you are talking about conversations...

Sorry for the hijack. Back to bookmarks now. 🤐

**Comment by der-lyse** [comment]
\<offtopic>
> But I bet they aren't hash replied that much.

I'm not so sure about that. It would be interesting to see an analysis on your crawled twtxt universe. :-)
\</offtopic>

In my opinion bookmarking is a client feature. Since the twtxt.net client runs in a browser, why not simply using the browser's bookmarks for that? It's decades old and works well, so why do we want to reinvent them?

What's the intent of making those bookmarks public? (twtxt is the only "social media thingy" I've ever used, so bare with me if this question sounds silly to you)

**Comment by sorenpeter** [comment]
I agreed that it should be a client feature.

Reason for having a bookmark.txt next to your twtxt.txt is to have all your data in one place. either on a pod or you own server or local disk. The bookmarks in the brower is for saving URLs of webpages. With twts we are talking about monologues  or parts of conversations.

The usecase for a bookmark.txt feature is to use it like you do in an email client or rss-reader (see above;) so you can get back to important pieces of information. Maybe you need to check something before posting a reply or you want to save a nice quote or link + who shared it.

If you got a bookmark.txt file with the suggested syntax it would be trivial to make it into its own twtxt-feed. A way of making a link-log or like-log. (not sure about the like-thing)

**Comment by prologic** [comment]
> In my opinion bookmarking is a client feature. Since the twtxt.net client runs in a browser, why not simply using the browser's bookmarks for that? It's decades old and works well, so why do we want to reinvent them?

So the only use-case I see for persisting user bookmarks on the `twtd` backend is say you want to be able to lookup a bookmark on Web or Mobile or even via the API.

Otherwise I agree with you, just use the browser's bookmarking feature, I've done so a few times myself.

> What's the intent of making those bookmarks public? (twtxt is the only "social media thingy" I've ever used, so bare with me if this question sounds silly to you)

Yeah I tend to agree, I'm not sure making one's own Twt bookmarks public is useful really. Plus I'd rather we keep the "plausible deniability" as much as possible (_hence twts are actively archived_)

**Comment by der-lyse** [comment]
> So the only use-case I see for persisting user bookmarks on the twtd backend is say you want to be able to lookup a bookmark on Web or Mobile or even via the API.

Oh, yes. Makes totally sense.
